src/RoomMembers.ts

```diff
@@ -0,0 +1,270 @@
+            // No need to record a leave for a user we didn't see joining.
+            return;
+        }
+        this._leaves.set(userId, timestamp);
+        this._joinsByUser.delete(userId);
+    }
+
+    /**
+     * Run a cleanup on the data structure.
+     */
+    public cleanup() {
+        if (this._leaves.size === 0) {
+            // Nothing to do.
+            return;
+        }
+        this._joinsByTimestamp = this._joinsByTimestamp.filter(join => this.isStillValid(join));
+        this._leaves = new Map();
+    }
+
+    /**
+     * Determine whether a `join` is still valid or has been superseded by a `leave`.
+     *
+     * @returns true if the `join` is still valid.
+     */
+    private isStillValid(join: Join): boolean {
+        const leaveTS = this._leaves.get(join.userId);
+        if (!leaveTS) {
+            // The user never left.
+            return true;
+        }
+        if (leaveTS > join.timestamp) {
+            // The user joined, then left, ignore this join.
+            return false;
+        }
+        // The user had left, but this is a more recent re-join.
+        return true;
+    }
+
+    /**
+     * Return a subset of the list of all the members, with their join date.
+     *
+     * @param since Only return members who have last joined at least as
+     * recently as `since`.
+     * @param max Only return at most `max` numbers.
+     * @returns A list of up to `max` members joined since `since`, ranked
+     * from most recent join to oldest join.
+     */
+    public members(since: Date, max: number): Join[] {
+        const result = [];
+        const ts = since.getTime();
+        // Spurious joins are legal, let's deduplicate them.
+        const users = new Set();
+        for (let i = this._joinsByTimestamp.length - 1; i >= 0; --i) {
+            if (result.length > max) {
+                // We have enough entries, let's return immediately.
+                return result;
+            }
+            const join = this._joinsByTimestamp[i];
+            if (join.timestamp < ts) {
+                // We have reached an older entry, everything will be `< since`,
+                // we won't find any other join to return.
+                return result;
+            }
+            if (this.isStillValid(join) && !users.has(join.userId)) {
+                // This entry is still valid, we'll need to return it.
+                result.push(join);
+                users.add(join.userId);
+            }
+        }
+        // We have reached the startup of Mjölnir.
+        return result;
+    }
+
+    /**
+     * Return the join date of a user.
+     *
+     * @returns a `Date` if the user is currently in the room and has joined
+     * since the start of Mjölnir, `null` otherwise.
+     */
+    public get(userId: string): Date | null {
+        let ts = this._joinsByUser.get(userId);
+        if (!ts) {
+            return null;
+        }
+        return new Date();
```

Comment:
shouldn't this be the `Date` that the user joined?

Reply:
Oops! Good catch!